---
layout: page
title: Distance and clustering lab
---

```{r options, echo=FALSE}
opts_chunk$set(fig.path=paste0("figure/", sub("(.*).Rmd","\\1",basename(knitr:::knit_concord$get('infile'))), "-"))
```

Let's load the tissue gene expression data which Rafa discussed in the lecture:

```{r}
# library(devtools)
# install_github("dagdata","genomicsclass")
library(dagdata)
data(tissuesGeneExpression)
library(Biobase)
rownames(tab) <- tab$filename
t <- ExpressionSet(e, AnnotatedDataFrame(tab))
t$Tissue <- factor(t$Tissue)
boxplot(exprs(t), range=0)
```

We will take the transpose of the vector, as the `dist` function will
compute distances between the rows, and we want sample-sample distances:
Note that the value in the 1,2 slot of the matrix of the object `d` is
the same as the square root of the sum of squared differences, i.e.,
the definition of the Euclidean distance.

```{r}
x <- t(exprs(t))
d <- dist(x)
class(d)
as.matrix(d)[1,2]
sqrt(sum((x[1,] - x[2,])^2))
```

## Clustering

We can perform hierarchical clustering based on the
distances defined above, using the `hclust` function.
The `plot` method will make a plot of the tree that results from
`hclust`. 

```{r}
hc <- hclust(d)
hc
plot(hc)
library(rafalib)
mypar()
plot(hc, labels=t$Tissue)
myplclust(hc, labels=t$Tissue, lab.col=as.fumeric(t$Tissue))
abline(h=120)
```

If we use the line above to cut the tree into clusters, we can examine
how the clusters overlap with the actual tissues:

```{r}
hclusters <- cutree(hc, h=120)
table(true=t$Tissue, cluster=hclusters)
```

We can also call the `kmeans` function to perform k-means clustering
as was explained in the lecture. Let's run k-means on the samples in
the space of the first two genes, so that we can observe the results
of the algorithm in the same space:

```{r}
plot(x[,1], x[,2])
km <- kmeans(x[,1:2], centers=3)
names(km)
plot(x[,1], x[,2], col=km$cluster, pch=16)
```

Note that if we perform k-means clustering using all of the genes, the
results are not the same as before. Looking at the first two genes
doesn't explain why the clusters we see were generated. So we need to
use a different method to *see* these inter-samples distances.

```{r}
km <- kmeans(x, centers=3)
plot(x[,1], x[,2], col=km$cluster, pch=16)
```

Let's try one last time with larger 'k':

```{r}
km <- kmeans(x, centers=7)
table(true=t$Tissue, cluster=km$cluster)
```
